VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+
+## Reconstructing a Variant
+
+It is possible to recover a full Variant value using a recursive algorithm, where the initial call is to `construct_variant` with the top-level Variant group fields.
+
+```python
+def construct_variant(metadata, value, typed_value):
+    """Constructs a Variant from value and typed_value"""
+    if typed_value is not None:
+        if isinstance(typed_value, dict):
+            # this is a shredded object
+            object_fields = {
+                name: construct_variant(metadata, field.value, field.typed_value)
+                for (name, field) in typed_value
+            }
+
+            if value is not None:
+                # this is a partially shredded object
+                assert isinstance(value, VariantObject), "partially shredded value must be an object"
+                assert typed_value.keys().isdisjoint(value.keys()), "object keys must be disjoint"
+
+                # union the shredded fields and non-shredded fields
+                return VariantObject(metadata, object_fields).union(VariantObject(metadata, value))
+
+            else:
+                return VariantObject(metadata, object_fields)
+
+        elif isinstance(typed_value, list):
+            # this is a shredded array
+            assert value is None, "shredded array must not conflict with variant value"
+
+            elements = [
+                construct_variant(metadata, elem.value, elem.typed_value)
+                for elem in list(typed_value)
+            ]
+            return VariantArray(metadata, elements)
+
+        else:
+            # this is a shredded primitive
+            assert value is None, "shredded primitive must not conflict with variant value"
+
+            return primitive_to_variant(typed_value)
+
+    elif value is not None:
+        return Variant(metadata, value)
+
+    else:
+        # value is missing
+        return None
+
+def primitive_to_variant(typed_value):
+    if isinstance(typed_value, int):
+        return VariantInteger(typed_value)
+    elif isinstance(typed_value, str):
+        return VariantString(typed_value)
+    ...
 ```
 
-# Backward and forward compatibility
 
-Shredding is an optional feature of Variant, and readers must continue to be able to read a group containing only a `value` and `metadata` field.
+## Backward and forward compatibility
 
-Any fields in the same group as `typed_value`/`variant_value` that start with `_` (underscore) can be ignored.
-This is intended to allow future backwards-compatible extensions.
-In particular, the field names `_metadata_key_paths` and any name starting with `_spark` are reserved, and should not be used by other implementations.
-Any extra field names that do not start with an underscore should be assumed to be backwards incompatible, and readers should fail when reading such a schema.
+Shredding is an optional feature of Variant, and readers must continue to be able to read a group containing only `value` and `metadata` fields.
```

Comment:
At this point, isn't non-shredded just a special case of shredded with no `typed_value` in the top level struct? I think it's automatically backwards compatible.

Reply:
I think the only thing that isn't backwards-compatible is that `value` is `optional` rather than `required` if you're shredding. But yes, writers are not required to shred.

Reply:
Do we need to add a note that once shredded, a file must be read using this spec and the typed_value portion can not be ignored?